man/ibqueryerrors.8
.TH IBQUERYERRORS 8 "Dec 31, 2009" "OpenIB" "OpenIB Diagnostics"

.SH NAME
ibqueryerrors \- query and report non-zero IB port counters

.SH SYNOPSIS
.B ibqueryerrors
[-s <err1,err2,...> -c -r -C <ca_name> -P <ca_port> -G <node_guid>
-D <direct_route> -d -k -K \-\-load\-cache <filename>]

.SH DESCRIPTION
.PP
ibqueryerrors reports port counters.  This is similar to ibcheckerrors with
the additional ability to filter out selected errors, include the optional
transmit and receive data counters, and report full link information for the
link reported.

.SH OPTIONS

.PP
.TP
\fB\-s <err1,err2,...>\fR
Suppress the errors listed in the comma separated list provided.
.TP
\fB\-c\fR
Suppress some of the common "side effect" counters.  These counters usually do
not indicate an error condition and can be usually be safely ignored.
.TP
\fB\-G <node_guid>\fR
Report results only for the node guid specified.
.TP
\fB\-S <node_guid>\fR
\-S is provided only for backward compatibility and works the same as "-G"
.TP
\fB\-D <direct_route>\fR
Report results only for the switch specified by the direct route path.
.TP
\fB\-r\fR
Report the port information.  This includes LID, port, external port (if
applicable), link speed setting, remote GUID, remote port, remote external port
(if applicable), and remote node description information.
.TP
\fB\-\-data\fR
Include the optional transmit and receive data counters.
.TP
\fB\-\-switch\fR  print data for switches only
.TP
\fB\-\-ca\fR  print data for CA's only
.TP
\fB\-\-router\fR  print data for routers only
.TP
\fB\-\-clear\-errors\fR \fB\-k\fR Clear error counters after read.
\-k and \-K can be used together to clear both errors and counters.
.TP
\fB\-\-clear\-counts\fR \fB\-K\fR Clear data counters after read.
\fBCAUTION\fR clearing data counters will occur regardless of if they are
printed or not.  This is because data counters are only \fBprinted\fR on ports
which have errors.  This means if a port has 0 errors and the \-K option is
specified the data counters will be cleared without any printed output.
.TP
\fB\-\-details\fR include receive error and transmit discard details
.TP
\fB\-\-load\-cache\fR <filename>
Load and use the cached ibnetdiscover data stored in the specified
filename.  May be useful for outputting and learning about other
fabrics or a previous state of a fabric.  Cannot be used if user
specifies a directo route path.  See
.B ibnetdiscover
for information on caching ibnetdiscover output.
.TP
\fB\-R\fR  (This option is obsolete and does nothing)

.SH COMMON OPTIONS
.PP
\-d      raise the IB debugging level.
	May be used several times (-ddd or -d -d -d).
.PP
\-e      show send and receive errors (timeouts and others)
.PP
\-h      show the usage message
.PP
\-v      increase the application verbosity level.
	May be used several times (-vv or -v -v -v)
.PP
\-V      show the version info.

# Other common flags:
.PP
\-C <ca_name>    use the specified ca_name.
.PP
\-P <ca_port>    use the specified ca_port.
.PP
\-t <timeout_ms> override the default timeout for the solicited mads.


.SH AUTHOR
.TP
Ira Weiny
.RI < weiny2@llnl.gov >


ChangeLog
commit c3cc42fffb2d1266cb2c3389a5fa3a8fd4f3f9e9
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:52:23 2009 -0500

    Spin 1.2.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit bfdca7d4392aaee9d8253d306378b036b8a60874
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:52:22 2009 -0500

    Remove immediate work requests
    
    The aren't supported/defined in iWARP.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 7b7460e47ad82ea81101230642c43ee10c6ad517
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:18:12 2009 -0500

    Handle 0B IO better.
    
    Allow num_sge == 0 to indicate a 0B IO.
    
    Also, for read, if length is 0 then create a valid 0B read request
    regardless of the user's specified lkey/rkey.  This avoids 0 stag
    connection terminations.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit e901d7d2ab0e9429118765e98e69a3992b91c584
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:00:54 2009 -0500

    Reset the T3 wq when the qp moves to RESET.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0f71e2dc0e1a40f4954f30f14deae965818c3c4e
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:10:57 2009 -0500

    Spin release 1.2.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 76d2a6347ab753d26237dc94c95da8d0c6c0c251
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:07:44 2009 -0500

    Support for doorbell overflow avoidance.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ad77c21cafde669c878032fc0f46635408568498
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:07:20 2009 -0500

    Support sq_sig_all QP attribute.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f6ea9eb207fa4b68f3a9c6f016bc8a262edb1808
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:58:03 2009 -0500

    Spin release version 1.2.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0ec532c2b6b92be6adf58849c9e3d215a50d306d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:52:51 2009 -0500

    Support new device ids: 0x35, 0x36.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f685c8fe7e77e64614d825e563dd9f02a0b1ae16
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:46 2008 -0500

    Spin release 1.2.2
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ebd19d077d331270805aefa9ca00d3661561ba87
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:27 2008 -0500

    Relax the required firmware level.
    
    Turns out that libcxgb3 can really run on either 6.0 or 5.0
    firmware.  So I'm relaxing the requirement...
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 6f7485feb244d8571fcab2292ef92c97bea48df0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue May 13 09:09:35 2008 -0500

    Spin release 1.2.1.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 309916b23ba384a0bd297e7e51b6b89bff357a78
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:45:01 2008 -0500

    Use local stag == 1 to silently drop the p2p read response.
    
    The current code fails if there are read completions -and-
    the p2p read response in the cq when its flushed.
    
    This fix uses the fact that the RTR read is to target stag 1,
    which is not a valid stag.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 67d519c4a9e28e54a9e436404156c9ae7eb60314
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:25:41 2008 -0500

    Handle wrapping when flushing the software sq.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a3dcea751a94ad5e0a9cad6c55a839a9848f327c
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Spin 1.2.0 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit cc2ef741842883233a77e18d65a828c711e581a3
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit 92322b93f48fa98b4fadb736ae49dafda237272a
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Apr 24 14:23:50 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit a8fb6221af4d914f39c1e71f9331f6d428d698c2
Author: Steve Wise (Open Grid Computing) <swise@hosting.openfabrics.org>
Date:   Tue Apr 1 15:30:11 2008 -0700

    Spin 1.1.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit db41ebc10f4cc4c2fe7826589636ab7d4b804218
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 26 08:27:12 2008 -0600

    Provide an empty stub for iwch_query_qp.
    
    This prevents crashes when calling ibv_query_qp due to an expected
    library provided function.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ec4f7f1c6b8beceb9001519bc4e2ee89fbb2e6d0
Author: Dotan Barak <dotanb@dev.mellanox.co.il>
Date:   Tue Feb 26 08:23:36 2008 -0600

    Remove duplicated line.
    
    Signed-off-by: Dotan Barak <dotanb@dev.mellanox.co.il>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 8218482d03898f859e1fa272649fafc64f91f05f
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Feb 12 15:16:03 2008 -0600

    Spin release 1.1.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 1bc6930d8884646ef73db71a95cc32c96c20ba32
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 12 14:46:10 2008 -0600

    libcxgb3: zeroing of wc_flags
    

src/ibqueryerrors.c
/*
 * Copyright (c) 2004-2009 Voltaire Inc.  All rights reserved.
 * Copyright (c) 2007 Xsigo Systems Inc.  All rights reserved.
 * Copyright (c) 2008 Lawrence Livermore National Lab.  All rights reserved.
 * Copyright (c) 2009 HNR Consulting.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdarg.h>
#include <time.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <inttypes.h>

#include <complib/cl_nodenamemap.h>
#include <infiniband/ibnetdisc.h>
#include <infiniband/mad.h>

#include "ibdiag_common.h"

struct ibmad_port *ibmad_port;
static char *node_name_map_file = NULL;
static nn_map_t *node_name_map = NULL;
static char *load_cache_file = NULL;

int data_counters = 0;
int port_config = 0;
uint64_t node_guid = 0;
char *node_guid_str = NULL;
#define SUP_MAX 64
int sup_total = 0;
enum MAD_FIELDS suppressed_fields[SUP_MAX];
char *dr_path = NULL;
uint8_t node_type_to_print = 0;
unsigned clear_errors = 0, clear_counts = 0, details = 0;

#define PRINT_SWITCH 0x1
#define PRINT_CA     0x2
#define PRINT_ROUTER 0x4
#define PRINT_ALL 0xFF		/* all nodes default flag */

static unsigned int get_max(unsigned int num)
{
	unsigned r = 0;		// r will be lg(num)

	while (num >>= 1)	// unroll for more speed...
		r++;

	return (1 << r);
}

static void get_msg(char *width_msg, char *speed_msg, int msg_size,
		    ibnd_port_t * port)
{
	char buf[64];
	uint32_t max_speed = 0;

	uint32_t max_width = get_max(mad_get_field(port->info, 0,
						   IB_PORT_LINK_WIDTH_SUPPORTED_F)
				     & mad_get_field(port->remoteport->info, 0,
						     IB_PORT_LINK_WIDTH_SUPPORTED_F));
	if ((max_width & mad_get_field(port->info, 0,
				       IB_PORT_LINK_WIDTH_ACTIVE_F)) == 0)
		// we are not at the max supported width
		// print what we could be at.
		snprintf(width_msg, msg_size, "Could be %s",
			 mad_dump_val(IB_PORT_LINK_WIDTH_ACTIVE_F,
				      buf, 64, &max_width));

	max_speed = get_max(mad_get_field(port->info, 0,
					  IB_PORT_LINK_SPEED_SUPPORTED_F)
			    & mad_get_field(port->remoteport->info, 0,
					    IB_PORT_LINK_SPEED_SUPPORTED_F));
	if ((max_speed & mad_get_field(port->info, 0,
				       IB_PORT_LINK_SPEED_ACTIVE_F)) == 0)
		// we are not at the max supported speed
		// print what we could be at.
		snprintf(speed_msg, msg_size, "Could be %s",
			 mad_dump_val(IB_PORT_LINK_SPEED_ACTIVE_F,
				      buf, 64, &max_speed));
}

static void print_port_config(char *node_name, ibnd_node_t * node, int portnum)
{
	char width[64], speed[64], state[64], physstate[64];
	char remote_str[256];
	char link_str[256];
	char width_msg[256];
	char speed_msg[256];
	char ext_port_str[256];
	int iwidth, ispeed, istate, iphystate;

	ibnd_port_t *port = node->ports[portnum];

	if (!port)
		return;

	iwidth = mad_get_field(port->info, 0, IB_PORT_LINK_WIDTH_ACTIVE_F);
	ispeed = mad_get_field(port->info, 0, IB_PORT_LINK_SPEED_ACTIVE_F);
	istate = mad_get_field(port->info, 0, IB_PORT_STATE_F);
	iphystate = mad_get_field(port->info, 0, IB_PORT_PHYS_STATE_F);

	remote_str[0] = '\0';
	link_str[0] = '\0';
	width_msg[0] = '\0';
	speed_msg[0] = '\0';

	snprintf(link_str, 256, "(%3s %s %6s/%8s)",
		 mad_dump_val(IB_PORT_LINK_WIDTH_ACTIVE_F, width, 64, &iwidth),
		 mad_dump_val(IB_PORT_LINK_SPEED_ACTIVE_F, speed, 64, &ispeed),
		 mad_dump_val(IB_PORT_STATE_F, state, 64, &istate),
		 mad_dump_val(IB_PORT_PHYS_STATE_F, physstate, 64, &iphystate));

	if (port->remoteport) {
		char *rem_node_name = NULL;

		if (port->remoteport->ext_portnum)
			snprintf(ext_port_str, 256, "%d",
				 port->remoteport->ext_portnum);
		else
			ext_port_str[0] = '\0';

		get_msg(width_msg, speed_msg, 256, port);

		rem_node_name = remap_node_name(node_name_map,
						port->remoteport->node->guid,
						port->remoteport->node->
						nodedesc);

		snprintf(remote_str, 256,
			 "0x%016" PRIx64 " %6d %4d[%2s] \"%s\" (%s %s)\n",
			 port->remoteport->node->guid,
			 port->remoteport->base_lid ? port->remoteport->
			 base_lid : port->remoteport->node->smalid,
			 port->remoteport->portnum, ext_port_str, rem_node_name,
			 width_msg, speed_msg);

		free(rem_node_name);
	} else
		snprintf(remote_str, 256, "           [  ] \"\" ( )\n");

	if (port->ext_portnum)
		snprintf(ext_port_str, 256, "%d", port->ext_portnum);
	else
		ext_port_str[0] = '\0';

	if (node->type == IB_NODE_SWITCH)
		printf("       Link info: %6d", node->smalid);
	else
		printf("       Link info: %6d", port->base_lid);

	printf("%4d[%2s] ==%s==>  %s",
	       port->portnum, ext_port_str, link_str, remote_str);
}

static int suppress(enum MAD_FIELDS field)
{
	int i = 0;
	for (i = 0; i < sup_total; i++)
		if (field == suppressed_fields[i])
			return 1;
	return 0;
}

static void report_suppressed(void)
{
	int i = 0;
	printf("Suppressing:");
	for (i = 0; i < sup_total; i++)
		printf(" %s", mad_field_name(suppressed_fields[i]));
	printf("\n");
}

static int query_and_dump(char *buf, size_t size, ib_portid_t * portid,
			  ibnd_node_t * node, char *node_name, int portnum,
			  const char *attr_name, uint16_t attr_id,
			  int start_field, int end_field)
{
	uint8_t pc[1024];
	uint32_t val = 0;
	int i, n;

	memset(pc, 0, sizeof(pc));

	if (!pma_query_via(pc, portid, portnum, ibd_timeout, attr_id,
			   ibmad_port)) {
		IBWARN("%s query failed on %s, %s port %d", attr_name,
		       node_name, portid2str(portid), portnum);
		return 0;
	}

	for (n = 0, i = start_field; i < end_field; i++) {
		mad_decode_field(pc, i, (void *)&val);
		if (val)
			n += snprintf(buf + n, size - n, " [%s == %u]",
				      mad_field_name(i), val);
	}

	return n;
}

static void print_results(ib_portid_t * portid, char *node_name,
			  ibnd_node_t * node, uint8_t * pc, int portnum,
			  int *header_printed)
{
	char buf[1024];
	char *str = buf;
	uint32_t val = 0;
	int i, n;

	for (n = 0, i = IB_PC_ERR_SYM_F; i <= IB_PC_VL15_DROPPED_F; i++) {
		if (suppress(i))
			continue;

		/* this is not a counter, skip it */
		if (i == IB_PC_COUNTER_SELECT2_F)
			continue;

		mad_decode_field(pc, i, (void *)&val);
		if (val)
			n += snprintf(str + n, 1024 - n, " [%s == %u]",
				      mad_field_name(i), val);

		/* If there are PortXmitDiscards, get details (if supported) */
		if (i == IB_PC_XMT_DISCARDS_F && details && val) {
			n += query_and_dump(str + n, sizeof(buf) - n, portid,
					    node, node_name, portnum,
					    "PortXmitDiscardDetails",
					    IB_GSI_PORT_XMIT_DISCARD_DETAILS,
					    IB_PC_RCV_LOCAL_PHY_ERR_F,
					    IB_PC_RCV_ERR_LAST_F);
			/* If there are PortRcvErrors, get details (if supported) */
		} else if (i == IB_PC_ERR_RCV_F && details && val) {
			n += query_and_dump(str + n, sizeof(buf) - n, portid,
					    node, node_name, portnum,
					    "PortRcvErrorDetails",
					    IB_GSI_PORT_RCV_ERROR_DETAILS,
					    IB_PC_XMT_INACT_DISC_F,
					    IB_PC_XMT_DISC_LAST_F);
		}
	}

	if (!suppress(IB_PC_XMT_WAIT_F)) {
		mad_decode_field(pc, IB_PC_XMT_WAIT_F, (void *)&val);
		if (val)
			n += snprintf(str + n, 1024 - n, " [%s == %u]",
				      mad_field_name(IB_PC_XMT_WAIT_F), val);
	}

	/* if we found errors. */
	if (n != 0) {
		if (data_counters)
			for (i = IB_PC_XMT_BYTES_F; i <= IB_PC_RCV_PKTS_F; i++) {
				uint64_t val64 = 0;
				mad_decode_field(pc, i, (void *)&val64);
				if (val64)
					n += snprintf(str + n, 1024 - n,
						      " [%s == %" PRIu64 "]",
						      mad_field_name(i), val64);
			}

		if (!*header_printed) {
			printf("Errors for 0x%" PRIx64 " \"%s\"\n", node->guid,
			       node_name);
			*header_printed = 1;
		}

		printf("   GUID 0x%" PRIx64 " port %d:%s\n", node->guid,
		       portnum, str);
		if (port_config)
			print_port_config(node_name, node, portnum);
	}
}

static int query_cap_mask(ib_portid_t * portid, char *node_name, int portnum,
			  uint16_t * cap_mask)
{
	uint8_t pc[1024];
	uint16_t rc_cap_mask;

	/* PerfMgt ClassPortInfo is a required attribute */
	if (!pma_query_via(pc, portid, portnum, ibd_timeout, CLASS_PORT_INFO,
			   ibmad_port)) {
		IBWARN("classportinfo query failed on %s, %s port %d",
		       node_name, portid2str(portid), portnum);
		return -1;
	}

	/* ClassPortInfo should be supported as part of libibmad */
	memcpy(&rc_cap_mask, pc + 2, sizeof(rc_cap_mask));	/* CapabilityMask */

	*cap_mask = ntohs(rc_cap_mask);
	return 0;
}

static void print_port(ib_portid_t * portid, uint16_t cap_mask, char *node_name,
		       ibnd_node_t * node, int portnum, int *header_printed)
{
	uint8_t pc[1024];

	memset(pc, 0, 1024);

	if (!pma_query_via(pc, portid, portnum, ibd_timeout,
			   IB_GSI_PORT_COUNTERS, ibmad_port)) {
		IBWARN("IB_GSI_PORT_COUNTERS query failed on %s, %s port %d",
		       node_name, portid2str(portid), portnum);
		return;
	}
	if (!(cap_mask & 0x1000)) {
		/* if PortCounters:PortXmitWait not supported clear this counter */
		uint32_t foo = 0;
		mad_encode_field(pc, IB_PC_XMT_WAIT_F, &foo);
	}
	print_results(portid, node_name, node, pc, portnum, header_printed);
}

static void clear_port(ib_portid_t * portid, uint16_t cap_mask,
		       char *node_name, int port)
{
	uint8_t pc[1024];
	/* bits defined in Table 228 PortCounters CounterSelect and
	 * CounterSelect2
	 */
	uint32_t mask = 0;

	if (!clear_errors && !clear_counts)
		return;

	if (clear_errors) {
		mask |= 0xFFF;
		if (cap_mask & 0x1000)
			mask |= 0x10000;
	}
	if (clear_counts)
		mask |= 0xF000;

	if (!performance_reset_via(pc, portid, port, mask, ibd_timeout,
				   IB_GSI_PORT_COUNTERS, ibmad_port))
		IBERROR("Failed to reset errors %s port %d", node_name, port);

	if (details && clear_errors) {
		performance_reset_via(pc, portid, port, 0xf, ibd_timeout,
				      IB_GSI_PORT_XMIT_DISCARD_DETAILS,
				      ibmad_port);
		performance_reset_via(pc, portid, port, 0x3f, ibd_timeout,
				      IB_GSI_PORT_RCV_ERROR_DETAILS,
				      ibmad_port);
	}
}

void print_node(ibnd_node_t * node, void *user_data)
{
	int header_printed = 0;
	int p = 0;
	int startport = 1;
	int type = 0;
	int all_port_sup = 0;
	ib_portid_t portid = { 0 };
	uint16_t cap_mask = 0;
	char *node_name = NULL;

	switch (node->type) {
	case IB_NODE_SWITCH:
		type = PRINT_SWITCH;
		break;
	case IB_NODE_CA:
		type = PRINT_CA;
		break;
	case IB_NODE_ROUTER:
		type = PRINT_ROUTER;
		break;
	}

	if ((type & node_type_to_print) == 0)
		return;

	if (node->type == IB_NODE_SWITCH && node->smaenhsp0)
		startport = 0;

	node_name = remap_node_name(node_name_map, node->guid, node->nodedesc);

	for (p = startport; p <= node->numports; p++) {
		if (node->ports[p]) {
			if (node->type == IB_NODE_SWITCH)
				ib_portid_set(&portid, node->smalid, 0, 0);
			else
				ib_portid_set(&portid, node->ports[p]->base_lid,
					      0, 0);

			if (query_cap_mask(&portid, node_name, p, &cap_mask) <
			    0)
				continue;

			if (cap_mask & 0x100)
				all_port_sup = 1;

			print_port(&portid, cap_mask, node_name, node, p,
				   &header_printed);
			if (!all_port_sup)
				clear_port(&portid, cap_mask, node_name, p);
		}
	}

	if (all_port_sup)
		clear_port(&portid, cap_mask, node_name, 0xFF);

	free(node_name);
}

static void add_suppressed(enum MAD_FIELDS field)
{
	if (sup_total >= SUP_MAX) {
		IBWARN("Maximum (%d) fields have been suppressed; skipping %s",
		       sup_total, mad_field_name(field));
		return;
	}
	suppressed_fields[sup_total++] = field;
}

static void calculate_suppressed_fields(char *str)
{
	enum MAD_FIELDS f;
	char *val, *lasts = NULL;
	char *tmp = strdup(str);

	val = strtok_r(tmp, ",", &lasts);
	while (val) {
		for (f = IB_PC_FIRST_F; f <= IB_PC_LAST_F; f++)
			if (strcmp(val, mad_field_name(f)) == 0)
				add_suppressed(f);
		val = strtok_r(NULL, ",", &lasts);
	}

	free(tmp);
}

static int process_opt(void *context, int ch, char *optarg)
{
	switch (ch) {
	case 's':
		calculate_suppressed_fields(optarg);
		break;
	case 'c':
		/* Right now this is the only "common" error */
		add_suppressed(IB_PC_ERR_SWITCH_REL_F);
		break;
	case 1:
		node_name_map_file = strdup(optarg);
		break;
	case 2:
		data_counters++;
		break;
	case 3:
		node_type_to_print |= PRINT_SWITCH;
		break;
	case 4:
		node_type_to_print |= PRINT_CA;
		break;
	case 5:
		node_type_to_print |= PRINT_ROUTER;
		break;
	case 6:
		details = 1;
		break;
	case 7:
		load_cache_file = strdup(optarg);
		break;
	case 'G':
	case 'S':
		node_guid_str = optarg;
		node_guid = strtoull(optarg, 0, 0);
		break;
	case 'D':
		dr_path = strdup(optarg);
		break;
	case 'r':
		port_config++;
		break;
	case 'R':		/* nop */
		break;
	case 'k':
		clear_errors = 1;
		break;
	case 'K':
		clear_counts = 1;
		break;
	default:
		return -1;
	}

	return 0;
}

int main(int argc, char **argv)
{
	int resolved = -1;
	ib_portid_t portid = { 0 };
	int rc = 0;
	ibnd_fabric_t *fabric = NULL;

	int mgmt_classes[4] = { IB_SMI_CLASS, IB_SMI_DIRECT_CLASS, IB_SA_CLASS,
		IB_PERFORMANCE_CLASS
	};

	const struct ibdiag_opt opts[] = {
		{"suppress", 's', 1, "<err1,err2,...>",
		 "suppress errors listed"},
		{"suppress-common", 'c', 0, NULL,
		 "suppress some of the common counters"},
		{"node-name-map", 1, 1, "<file>", "node name map file"},
		{"node-guid", 'G', 1, "<node_guid>", "query only <node_guid>"},
		{"", 'S', 1, "<node_guid>",
		 "Same as \"-G\" for backward compatibility"},
		{"Direct", 'D', 1, "<dr_path>",
		 "query only switch specified by <dr_path>"},
		{"report-port", 'r', 0, NULL,
		 "report port configuration information"},
		{"GNDN", 'R', 0, NULL,
		 "(This option is obsolete and does nothing)"},
		{"data", 2, 0, NULL, "include the data counters in the output"},
		{"switch", 3, 0, NULL, "print data for switches only"},
		{"ca", 4, 0, NULL, "print data for CA's only"},
		{"router", 5, 0, NULL, "print data for routers only"},
		{"details", 6, 0, NULL, "include transmit discard details"},
		{"clear-errors", 'k', 0, NULL,
		 "Clear error counters after read"},
		{"clear-counts", 'K', 0, NULL,
		 "Clear data counters after read"},
		{"load-cache", 7, 1, "<file>",
		 "filename of ibnetdiscover cache to load"},
		{0}
	};
	char usage_args[] = "";

	memset(suppressed_fields, 0, sizeof suppressed_fields);
	ibdiag_process_opts(argc, argv, NULL, "scnSrRDGL", opts, process_opt,
			    usage_args, NULL);

	argc -= optind;
	argv += optind;

	if (!node_type_to_print)
		node_type_to_print = PRINT_ALL;

	if (ibverbose)
		ibnd_debug(1);

	ibmad_port = mad_rpc_open_port(ibd_ca, ibd_ca_port, mgmt_classes, 4);
	if (!ibmad_port)
		IBERROR("Failed to open port; %s:%d\n", ibd_ca, ibd_ca_port);

	if (ibd_timeout)
		mad_rpc_set_timeout(ibmad_port, ibd_timeout);

	node_name_map = open_node_name_map(node_name_map_file);

	if (dr_path && load_cache_file) {
		fprintf(stderr, "Cannot specify cache and direct route path\n");
		exit(1);
	}

	/* limit the scan the fabric around the target */
	if (dr_path) {
		if ((resolved =
		     ib_resolve_portid_str_via(&portid, dr_path, IB_DEST_DRPATH,
					       NULL, ibmad_port)) < 0)
			IBWARN("Failed to resolve %s; attempting full scan",
			       dr_path);
	} else if (node_guid_str) {
		if ((resolved =
		     ib_resolve_portid_str_via(&portid, node_guid_str,
					       IB_DEST_GUID, ibd_sm_id,
					       ibmad_port)) < 0)
			IBWARN("Failed to resolve %s; attempting full scan",
			       node_guid_str);
	}

	if (load_cache_file) {
		if ((fabric = ibnd_load_fabric(load_cache_file, 0)) == NULL) {
			fprintf(stderr, "loading cached fabric failed\n");
			exit(1);
		}
	} else {
		if (resolved >= 0 &&
		    !(fabric = ibnd_discover_fabric(ibmad_port, &portid, 0)))
			IBWARN("Single node discover failed;"
			       " attempting full scan");

		if (!fabric &&
		    !(fabric = ibnd_discover_fabric(ibmad_port, NULL, -1))) {
			fprintf(stderr, "discover failed\n");
			rc = 1;
			goto close_port;
		}
	}

	report_suppressed();

	if (node_guid_str) {
		ibnd_node_t *node = ibnd_find_node_guid(fabric, node_guid);
		if (node)
			print_node(node, NULL);
		else
			fprintf(stderr, "Failed to find node: %s\n",
				node_guid_str);
	} else if (dr_path) {
		ibnd_node_t *node = ibnd_find_node_dr(fabric, dr_path);
		uint8_t ni[IB_SMP_DATA_SIZE];

		if (!smp_query_via(ni, &portid, IB_ATTR_NODE_INFO, 0,
				   ibd_timeout, ibmad_port)) {
			rc = -1;
			goto destroy_fabric;
		}
		mad_decode_field(ni, IB_NODE_GUID_F, &(node_guid));

		node = ibnd_find_node_guid(fabric, node_guid);
		if (node)
			print_node(node, NULL);
		else
			fprintf(stderr, "Failed to find node: %s\n", dr_path);
	} else
		ibnd_iter_nodes(fabric, print_node, NULL);

destroy_fabric:
	ibnd_destroy_fabric(fabric);

close_port:
	mad_rpc_close_port(ibmad_port);
	close_node_name_map(node_name_map);
	exit(rc);
}
